File: Dynamic_Programming/115.Distinct-Subsequences/115.Distinct-Subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) 
    {
        int m = s.size();
        int n = t.size();
        auto dp = vector<vector<long long>>(m+1,vector<long long>(n+1,0));
        s = "#"+s;
        t = "#"+t;
                
        if (m<n) 
            return 0;
        if (m==n)
            return (s==t);
        
        for (int i=0; i<=m; i++)
            dp[i][0] = 1;
        
        for (int i=1; i<=m; i++)
            for (int j=1; j<=n; j++)
            {
                if (s[i]==t[j])
                    dp[i][j] = min(LLONG_MAX/2, dp[i-1][j] + dp[i-1][j-1]);
                else
                    dp[i][j] = dp[i-1][j];
            }
        return dp[m][n];
    }
};

File: Dynamic_Programming/115.Distinct-Subsequences/115.Distinct-Subsequences.py
class Solution:
    def numDistinct(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: int
        """
        M,N = len(s),len(t)
        s = "#"+s
        t = "#"+t
        dp = [[0 for _ in range(N+1)] for _ in range(M+1)]
        for i in range(M+1):
            dp[i][0] = 1
        for i in range(1,M+1):
            for j in range(1,N+1):                
                if s[i]==t[j]:
                    dp[i][j]+=dp[i-1][j-1]                
                dp[i][j]+=dp[i-1][j]
        return dp[M][N]
        
        
